### PR TITLE
include/uk/plat/memory.h fix for error #766

### LINKED ISSUE #766: lib/ukboot: Critical error when running `app-helloworld` on linuxu
### Describe the bug

When running the linuxu application for helloworld, it crashes due to a problem with memory allocator from `boot.c`

### Steps to reproduce

Run the [app-helloworld](https://github.com/unikraft/app-helloworld) application with Unikraft v0.12.

### Expected behavior

The program to dsplay the message "Hello world".

### Which architectures were you using or does this bug affect?

`linuxu`

### Which platforms were you using or does this bug affect?

`x86_64`

### Relevant log output

```shell
[ 2692.514525] Info: [libukboot] <boot.c @  245> Unikraft constructor table at 0x40f000 - 0x40f008
[ 2692.514558] Info: [libukboot] <boot.c @  264> Initialize memory allocator...
[ 2692.514571] CRIT: [libukboot] <boot.c @  268> Failed to initialize memory allocator
```

### PR DESCRIPTION
**Pull request by Daparrag**
### Prerequisite checklist

 - [X] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [X] Tested your changes against relevant architectures and platforms;
 - [X] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [X] Updated relevant documentation.

### Base target

 - Architecture(s): x86_64
 - Platform(s): linuxu
 - Application(s): [e.g. `helloword` or N/A]

### Additional configuration
none

### Description of changes
according to https://github.com/unikraft/unikraft/issues/766 the helloworld example was broken because an invalid memory allocator. 
~~~~
[443184.280410] Info: [libukboot] <boot.c @  245> Unikraft constructor table at 0x40e000 - 0x40e008
[443184.280459] dbg:  [libukboot] <boot.c @  249> Call constructor: 0x40d400())...
[443184.280509] dbg:  [libcontext] <ectx.c @   72> Load/store of extended CPU state: XSAVEOPT
[443184.280547] Info: [libukboot] <boot.c @  264> Initialize memory allocator...
[443184.280570] CRIT: [libukboot] <boot.c @  268> Failed to initialize memory allocator
[Inferior 1 (process 34651) exited with code 01]
~~~~

This problem occurred because you attempted to access a page without first determining the total number of available pages. To clarify further, the function ukplat_memregion_find_next requires the count of memory regions to be validated before calling the ukplat_memregion_find_next function to obtain the desired memory region.

For this reason I add a sanity check  count = ukplat_memregion_count(); that will avoid the program to advance if the memory region is greater or equal than memory region count. 

With this changes the hello world  works as expected

---output after fixing
~~~~
 ./helloworld_linuxu-x86_64
[447357.835182] dbg:  [libukboot] <boot.c @  249> Call constructor: 0x40c1f0())...
[447357.835227] dbg:  [libcontext] <ectx.c @   72> Load/store of extended CPU state: XSAVEOPT
[447357.835281] dbg:  [liblinuxuplat] <memory.c @   84> No initrd present.
[447357.835321] dbg:  [libukboot] <boot.c @  189> Trying 0x7f1c14674000-0x7f1c14a74000 0x00
[447357.835401] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14676000 - 7f1c14678000 (order 1)
[447357.835440] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14678000 - 7f1c14680000 (order 3)
[447357.835508] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14680000 - 7f1c14700000 (order 7)
[447357.835590] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14700000 - 7f1c14800000 (order 8)
[447357.835687] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14800000 - 7f1c14a00000 (order 9)
[447357.835810] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14a00000 - 7f1c14a40000 (order 6)
[447357.835865] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14a40000 - 7f1c14a60000 (order 5)
[447357.835920] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14a60000 - 7f1c14a70000 (order 4)
[447357.835995] dbg:  [libukallocbbuddy] <bbuddy.c @  471> 7f1c14674000: Add allocate unit 7f1c14a70000 - 7f1c14a74000 (order 2)
[447357.836074] dbg:  [libcontext] <tls.c @  173> tls_area_init: target: 0x7f1c14676020 (16 bytes)
[447357.836110] dbg:  [libcontext] <tls.c @  175> tls_area_init: copy (.tdata): 0 bytes
[447357.836163] dbg:  [libcontext] <tls.c @  177> tls_area_init: uninitialized (.tbss): 4 bytes
[447357.836203] dbg:  [libcontext] <tls.c @  179> tls_area_init: pad: 4 bytes
[447357.836257] dbg:  [libcontext] <tls.c @  181> tls_area_init: tcb: 8 bytes
[447357.836327] dbg:  [libcontext] <tls.c @  183> tls_area_init: tcb self ptr: 0x7f1c14676028
[447357.836370] dbg:  [libcontext] <tls.c @  204> (tls_area): 7f1c14676020  00 00 00 00 00 00 00 00  28 60 67 14 1c 7f 00 00  |........(`g.....|
[447357.836642] dbg:  [libuksched] <thread.c @  254> uk_thread 0x7f1c146770b0 (idle): ctx:0x7f1c146770b0, ectx:0x7f1c14a70040, tlsp:0x7f1c14a70028
[447357.836678] dbg:  [libcontext] <tls.c @  173> tls_area_init: target: 0x7f1c14a70020 (16 bytes)
[447357.836727] dbg:  [libcontext] <tls.c @  175> tls_area_init: copy (.tdata): 0 bytes
[447357.836775] dbg:  [libcontext] <tls.c @  177> tls_area_init: uninitialized (.tbss): 4 bytes
[447357.836818] dbg:  [libcontext] <tls.c @  179> tls_area_init: pad: 4 bytes
[447357.836840] dbg:  [libcontext] <tls.c @  181> tls_area_init: tcb: 8 bytes
[447357.836862] dbg:  [libcontext] <tls.c @  183> tls_area_init: tcb self ptr: 0x7f1c14a70028
[447357.836887] dbg:  [libcontext] <tls.c @  204> (tls_area): 7f1c14a70020  00 00 00 00 00 00 00 00  28 00 a7 14 1c 7f 00 00  |........(.......|
[447357.837259] dbg:  [libcontext] <ctx.c @  133> ukarch_ctx 0x7f1c146770b0: entry:0x40af80(7f1c14677018), sp:0x7f1c14a50018
[447357.837308] dbg:  [libuksched] <thread.c @  254> uk_thread 0x7f1c14a71018 (init): ctx:0x7f1c14a71018, ectx:0x7f1c14a71100, tlsp:0x7f1c14676028
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
      Epimetheus 0.12.0~4c7352c0-custom
Hello world!
Arguments:  "./helloworld_linuxu-x86_64"

      _
    c'_'o  .--'
    (| |)_/            [447357.837418] dbg:  [libposix_time] <time.c @   65> (in      _                ep((const struct timespec*) 0x7ffd00121bc0, (struct times    c'o'o  .--.
    (| |)_/            [447358.087798] dbg:  [libposix_time] <time.c @   65> (in      _                ep((const struct timespec*) 0x7ffd00121bc0, (struct times    c'_'o  .-.
    (| |)_/   `        [447358.338165] dbg:  [libposix_time] <time.c @   65> (in      _                ep((const struct timespec*) 0x7ffd00121bc0, (struct times
~~~~

### CODE REVIEW

**Comment by razvand** [comment]
Thanks for the PR, @Daparrag. @RaduNichita , @mogasergiu , please check it out. Also, @marcrittinghaus , take a look when you get the time.

File: include/uk/plat/memory.h [hunk 1]
```diff
@@ -132,7 +132,15 @@ ukplat_memregion_find_next(int i, __u32 type, __u32 flags, __u32 fmask,
...
 	struct ukplat_memregion_desc *desc;
 	__u32 stype, sflags;
-	int rc;
+	int rc, count;
+    	count = ukplat_memregion_count();
+
+	/* Sanity check before to get a memory region using ukplat_memregion_get
+	 * this prevent  Initialize memory allocator in platform linuxu 
+	 * */
+    	if(i >= count){
```
Comment (mogasergiu) [inline include/uk/plat/memory.h]: I believe this check is already done by `ukplat_memregion_get`, isn't it?

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/kvm/x86/memory.c#L26

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/kvm/arm/memory.c#L173

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/xen/memory.c#L143

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/linuxu/memory.c#L185

File: lib/ukboot/boot.c [hunk 2]
```diff
@@ -187,12 +187,12 @@ static struct uk_alloc *heap_init()
 	 */
 	ukplat_memregion_foreach(&md, UKPLAT_MEMRT_FREE, 0, 0) {
 		uk_pr_debug("Trying %p-%p 0x%02x %s\n",
-			    (void *)md->vbase, (void *)(md->vbase + md->len),
-			    md->flags,
+			    (void *)md->vbase, (void *)(md->vbase + md->len)
```
Comment (mogasergiu) [inline lib/ukboot/boot.c]: I think commas at the end of the line are more conventional.

**Review by mogasergiu — COMMENTED** [review]
I believe a more conventional commit message would fit better here . Do a `git log <directory with changes>` locally to see how the others look. For example, I would rephrase this as:
```
include/uk/plat: Add check for memory region descriptor index

Make sure that the value of the memory region descriptor index is not
bigger than the total count of memory region descriptors.

Signed-off-by: ...
```

You wrote `lib/ukboot/boot.c` in the PR title but from what I can tell the actual changes are in `include/uk/plat/memory.h`

File: include/uk/plat/memory.h (hunk 1 above)
Comment (Daparrag) [reply include/uk/plat/memory.h]: perhaps the problem is that  the function ukplat_memregion_count not only count the number of memregions. this function also initialize some heap and I dont see that call here: [unikraft/plat/linuxu/memory.c](https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/linuxu/memory.c#L185)
~~~~
if (!have_heap) {
		rc = __linuxu_plat_heap_init();
		have_heap = (rc == 0) ? 1 : 0;
	}

	if (!have_initrd) {
		rc = __linuxu_plat_initrd_init();
		have_initrd = (rc == 0) ? 1 : 0;
	}
~~~~

**Review by mogasergiu — COMMENTED** [review]
The reason why your fix seems to work is caused by the fact that `ukplat_memregion_count()` will end up calling `__linuxu_plat_heap_init()`, which will fill in the, initially empty, `_liblinuxuplat_opts` structure through `__linuxu_plat_heap_init()`. Therefore, this fix would have worked without the sanity check as well. The call to `ukplat_memregion_count` would have been enough. Therefore, the fix would be, I guess, to actually somehow end up calling `__linuxu_plat_heap_init()` before we actually get to enumerating memory regions

File: include/uk/plat/memory.h (hunk 1 above)
Comment (mogasergiu) [reply include/uk/plat/memory.h]: I somehow just now received your reply, yes, indeed that is why the fix works. Not because of the sanity check.

Comment (Daparrag) [reply include/uk/plat/memory.h]: sorry, I also get confused. and yes, you are correct.
would it work if  add the call  ukplat_memregion_count as in [unikraft/plat/xen/memory.c](https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/xen/memory.c#L143)?

I can work in the changes, So, please let me know what would be the best way to do it. 

Thank you.

Comment (mogasergiu) [reply include/uk/plat/memory.h]: Yeah, that would look good I guess. But it would be a good idea, in my opinion, to also add a check at the beginning of `ukplat_memregion_count()` for the `static` variables, to avoid re-initialization attempts if one of them is 0. This is how I would see the function, since if you successfully went through it once, I guess you do not need to repeat the run (I might be wrong here, @marcrittinghaus might have more context).
```C
int ukplat_memregion_count(void)
{
	static int init = 0;
	int have_initrd = 0;
	int have_heap = 0;
	int rc = 0;

	if (init)
		return init;
	/*
	 * NOTE: The heap size and initrd file can be changed by a
	 * library parameter. We assume that those ones are processed
	 * by the boot library shortly before memory regions are
	 * scanned. This is why we initialize the heap here.
	 */
	rc = __linuxu_plat_heap_init();
	have_heap = (rc == 0) ? 1 : 0;

	rc = __linuxu_plat_initrd_init();
	have_initrd = (rc == 0) ? 1 : 0;

	init = have_heap + have_initrd;

	return init;
}
```

Comment (mogasergiu) [reply include/uk/plat/memory.h]: I say let's wait for the others' opinion as well before making any change :).

Comment (Daparrag) [reply include/uk/plat/memory.h]: Sure 😊

Comment (RaduNichita) [reply include/uk/plat/memory.h]: I agree with @mogasergiu about adding the check

**Review by RaduNichita — CHANGES_REQUESTED** [review]
It works, I tested on linuxu for x86_64.

Please fix the checkpatch warnings, besides what @mogasergiu already suggested about the commit message. 

I would also change the PR title to be more suggestive.

**Comment by Daparrag** [comment]
Good I will start with the modifications
Thanks

**Comment by Daparrag** [comment]
I added the modifications requested and open to any feedback.

Thank you for the support.

**Comment by Daparrag** [comment]
Hi @RaduNichita  @mogasergiu 

I add the new changes. However I still have errors reported by your boot. 

- concourse-ci/app-helloworld-staging/arm-linuxu/build — Failed to build with helloworld
- concourse-ci/checkpatch — Failed against checkpatch— Failed against checkpatch

I dont really understand what would be the issue. Can you please give me some details. 

Best,

**Comment by Daparrag** [comment]
I migrate this pull request to #797
